### PR TITLE
Ret 1935 - Task List 2: Respondent Details - Next layer - Save data & Final checks

### PR DESCRIPTION
**Pull request by lvo700**
### JIRA link

https://tools.hmcts.net/jira/browse/RET-1935

### Change description

Save data & Final checks for Task List 2: Respondent Details

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by JustMehmet — DISMISSED** [review]

**Review by ikirsanov — APPROVED** [review]

File: src/main/definitions/complexTypes/addressUK.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 export interface AddressUK {
   AddressLine1: string;
   AddressLine2: string;
-  AddressLine3: string;
-  County: string;
+  AddressLine3?: string;
```
Comment (marcjudge) [inline src/main/definitions/complexTypes/addressUK.ts:4]: Do these need to be optional still?

File: src/main/definitions/complexTypes/respondent.ts [hunk 2]
```diff
@@ -1,3 +1,12 @@
+import { NoAcasNumberReason, YesOrNo } from '../../definitions/case';
+
+import { Et1Address } from './et1Address';
+
 export interface RespondentType {
   respondent_name?: string;
+  respondent_address?: Et1Address;
+  respondent_ACAS_question: YesOrNo;
```
Comment (marcjudge) [inline src/main/definitions/complexTypes/respondent.ts:8]: I'd make all of these optional

File: src/main/definitions/complexTypes/addressUK.ts (hunk 1 above)
Comment (lvo700) [reply src/main/definitions/complexTypes/addressUK.ts]: No. Fixed.

File: src/main/definitions/complexTypes/respondent.ts (hunk 2 above)
Comment (lvo700) [reply src/main/definitions/complexTypes/respondent.ts]: Done

**Review by ikirsanov — DISMISSED** [review]

**Review by marcjudge — DISMISSED** [review]

File: src/main/components/form/validator.ts [hunk 3]
```diff
@@ -13,7 +13,7 @@ export const isFieldFilledIn: Validator = value => {
 export const isRespondentNameValid: Validator = value => {
   if (!value || (value as string).trim().length === 0) {
     return 'required';
-  } else if (!/(=?^.{5,100}$)/.test(value as string)) {
+  } else if (!/(=?^.{1,100}$)/.test(value as string)) {
```
Comment (marcjudge) [inline src/main/components/form/validator.ts:16]: This change doesn't seem to be covered by tests. Validator should allow minimum two characters

Comment (lvo700) [reply src/main/components/form/validator.ts:16]: Respondent name can be 1 character, see comments https://tools.hmcts.net/jira/browse/RET-1827

File: src/main/controllers/RespondentNameController.ts [hunk 4]
```diff
@@ -42,11 +41,12 @@ export default class RespondentNameController {
   }
 
   public post = (req: AppRequest, res: Response): void => {
+    setUserCaseForRespondent(req, this.form);
     const { saveForLater } = req.body;
     if (saveForLater) {
-      handleSaveAsDraft(res);
+      handleSessionErrors(req, res, this.form, PageUrls.CLAIM_SAVED);
+      handleUpdateDraftCase(req, this.logger);
```
Comment (JackReevies) [inline src/main/controllers/RespondentNameController.ts:48]: This line is the same in both branches. Is it worth moving this outside of the if statement entirely?

File: src/main/controllers/AcasCertNumController.ts [hunk 5]
```diff
@@ -64,13 +64,14 @@ export default class AcasCertNumController {
   }
 
   public post = (req: AppRequest, res: Response): void => {
+    setUserCaseForRespondent(req, this.form);
     const { saveForLater } = req.body;
 
     if (saveForLater) {
-      handleSaveAsDraft(res);
+      handleSessionErrors(req, res, this.form, PageUrls.CLAIM_SAVED);
+      handleUpdateDraftCase(req, this.logger);
```
Comment (JackReevies) [inline src/main/controllers/AcasCertNumController.ts:72]: This line is the same in both branches. Is it worth moving this outside of the if statement entirely?

File: src/main/controllers/NoAcasNumberController.ts [hunk 6]
```diff
@@ -71,11 +70,13 @@ export default class NoAcasNumberController {
   }
 
   public post = (req: AppRequest, res: Response): void => {
+    setUserCaseForRespondent(req, this.form);
     const { saveForLater } = req.body;
+
     if (saveForLater) {
-      handleSaveAsDraft(res);
+      handleSessionErrors(req, res, this.form, PageUrls.CLAIM_SAVED);
+      handleUpdateDraftCase(req, this.logger);
```
Comment (JackReevies) [inline src/main/controllers/NoAcasNumberController.ts:78]: This line is the same in both branches. Is it worth moving this outside of the if statement entirely?

File: src/main/controllers/WorkAddressController.ts [hunk 7]
```diff
@@ -34,10 +34,12 @@ export default class WorkAddressController {
   }
 
   public post = (req: AppRequest, res: Response): void => {
+    setUserCase(req, this.form);
     const { saveForLater } = req.body;
 
     if (saveForLater) {
-      handleSaveAsDraft(res);
+      handleSessionErrors(req, res, this.form, PageUrls.CLAIM_SAVED);
+      handleUpdateDraftCase(req, this.logger);
```
Comment (JackReevies) [inline src/main/controllers/WorkAddressController.ts:42]: This line is the same in both branches. Is it worth moving this outside of the if statement entirely?

File: src/main/controllers/AddressDetailsController.ts [hunk 8]
```diff
@@ -38,6 +39,7 @@ export default class AddressDetailsController {
         labelSize: null,
         attributes: {
           autocomplete: 'address-line2',
+          maxLength: 50,
         },
       },
```
Comment (JackReevies) [inline src/main/controllers/AddressDetailsController.ts:44]: Looks like theres a missing validation function here

File: src/main/controllers/RespondentAddressController.ts [hunk 9]
```diff
@@ -39,7 +40,7 @@ export default class RespondentAddressController {
         labelSize: null,
         attributes: {
           autocomplete: 'address-line2',
-          maxLength: 100,
+          maxLength: 50,
         },
       },
```
Comment (JackReevies) [inline src/main/controllers/RespondentAddressController.ts:45]: same here with missing validator function

File: src/main/controllers/RespondentNameController.ts (hunk 4 above)
Comment (lvo700) [reply src/main/controllers/RespondentNameController.ts]: Moved outside if statement

File: src/main/controllers/AcasCertNumController.ts (hunk 5 above)
Comment (lvo700) [reply src/main/controllers/AcasCertNumController.ts]: Moved outside if statement

File: src/main/controllers/NoAcasNumberController.ts (hunk 6 above)
Comment (lvo700) [reply src/main/controllers/NoAcasNumberController.ts]: Moved outside if statement

File: src/main/controllers/AddressDetailsController.ts (hunk 8 above)
Comment (lvo700) [reply src/main/controllers/AddressDetailsController.ts:45]: Validation added

File: src/main/controllers/RespondentAddressController.ts (hunk 9 above)
Comment (lvo700) [reply src/main/controllers/RespondentAddressController.ts:46]: Validation added

**Review by marcjudge — APPROVED** [review]

**Pull request merged**